### PR TITLE
Option to not auto-create index during expand_dims

### PR DESCRIPTION
**Pull request by TomNicholas**
- [x] Solves part of #8871 by pulling out part of https://github.com/pydata/xarray/pull/8872#issuecomment-2027571714 
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] ~~New functions/methods are listed in `api.rst`~~

TODO:
- [x] Add new kwarg to `DataArray.expand_dims`
- [ ] Add examples to docstrings?
- [x] Check it actually solves the problem in #8872

### CODE REVIEW

File: xarray/tests/test_dataset.py [hunk 1]
```diff
@@ -3431,6 +3431,36 @@ def test_expand_dims_kwargs_python36plus(self) -> None:
...
 
+    @pytest.mark.parametrize("create_1d_index_flag", [True, False])
+    def test_expand_dims_create_index_data_variable(self, create_1d_index_flag):
+        # data variables should not gain an index ever
+        ds = Dataset({"x": 0})
+        expanded = ds.expand_dims("x", create_1d_index=create_1d_index_flag)
+
+        # TODO I can't just create the expected dataset directly using constructor because of GH issue 8959
+        # expected = Dataset(data_vars={"x": ("x", [0])})
+        expected = Dataset({"x": ("x", [0])}).drop_indexes("x").reset_coords("x")
```
Comment (TomNicholas) [inline xarray/tests/test_dataset.py:3442]: See #8959

Comment (dcherian) [reply xarray/tests/test_dataset.py:3443]: I didn't think we should merge without either updating the invariants or fixing the behavior or test

Comment (TomNicholas) [reply xarray/tests/test_dataset.py:3443]: Yeah I agree, I wouldn't really have felt comfortable merging this at this point either.

Comment (TomNicholas) [reply xarray/tests/test_dataset.py:3443]: So the action here is to
1) Use the [private version of the asserts](https://github.com/pydata/xarray/blob/b0036749542145794244dee4c4869f3750ff2dee/xarray/tests/__init__.py#L286), which have a `check_default_indexes` kwarg
2) We should probably publicly expose that kwarg in the public assert functions API
3) We should write a little documentation explaining what it does, to go in the [docs page on testing](https://docs.xarray.dev/en/stable/user-guide/testing.html#testing-your-code)

Comment (dcherian) [reply xarray/tests/test_dataset.py:3443]: The private version *should* be getting used in all the tests anyway.

I don't know about exposing it publicly. It's our job to make sure they're always satisfied when using the API.

Comment (TomNicholas) [reply xarray/tests/test_dataset.py:3443]: > The private version should be getting used in all the tests anyway.

+1

> I don't know about exposing it publicly. It's our job to make sure they're always satisfied when using the API.

Isn't that only true if indexes are a hidden private part of the data model? If they are public, users should have control over them (including whether or not they exist for any given 1D dimension coordinate).

Perhaps we should not be exposing the kwarg, but instead setting `check_default_indexes=False` on the public asserts in favour of changing them to actually compare all the indexes and report if the indexes are unequal. i.e. what #5812 is suggesting.

Comment (dcherian) [reply xarray/tests/test_dataset.py:3443]: > hanging them to actually compare all the indexes and report if the indexes are unequal.

This makes sense to me.

Comment (TomNicholas) [reply xarray/tests/test_dataset.py]: Fixed in https://github.com/pydata/xarray/pull/8960/commits/62e750f46c55209b904cc14ece174fa58cce546a by using private versions of assertions. The rest of this discussion is already covered by #5812.

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -2557,6 +2557,7 @@ def expand_dims(
         self,
         dim: None | Hashable | Sequence[Hashable] | Mapping[Any, Any] = None,
         axis: None | int | Sequence[int] = None,
+        create_1d_index: bool = True,
```
[the remark this answers is not in the document]
Comment (TomNicholas) [reply xarray/core/dataarray.py]: Fixed in https://github.com/pydata/xarray/pull/8960/commits/f86c82f1c9efdf0b207f1fb537aea2cefe287c82

**Comment by TomNicholas** [comment]
@dcherian I've addressed your comments, so would like to merge this.

**Comment by dcherian** [comment]
[also good opportunity to extend our stateful tests in a future PR if you want to dip your toes there ;) ]

File: doc/whats-new.rst [hunk 3]
```diff
@@ -29,6 +29,9 @@ New Features
   for example, will retain the object.  However, one cannot do operations that are not possible on the `ExtensionArray`
   then, such as broadcasting.
   By `Ilan Gold <https://github.com/ilan-gold>`_.
+- Added the option to avoid automatically creating 1D pandas indexes in :py:meth:`Dataset.expand_dims()`, by passing the new kwarg
+  `create_1d_index=False`. (:pull:`8960`)
```
[the remark this answers is not in the document]
Comment (TomNicholas) [reply doc/whats-new.rst:33]: Agh my find-replace missed the docs

File: xarray/tests/test_dataset.py [hunk 4]
```diff
@@ -3431,6 +3431,47 @@ def test_expand_dims_kwargs_python36plus(self) -> None:
         )
         assert_identical(other_way_expected, other_way)
 
+    @pytest.mark.parametrize("create_index_flag", [True, False])
+    def test_expand_dims_create_index_data_variable(self, create_index_flag):
+        # data variables should not gain an index ever
+        ds = Dataset({"x": 0})
+        expanded = ds.expand_dims("x", create_index=create_index_flag)
```
Comment (dcherian) [inline xarray/tests/test_dataset.py:3438]: Shouldn't this raise a warning for `create_index=True` since that option is effectively ignored.

Comment (TomNicholas) [reply xarray/tests/test_dataset.py:3438]: Why would that raise a warning? You think a kwarg that is irrelevant when passed should raise a warning? 

I also feel like that could create unnecessary warnings, especially if we made a global `create_index` flag later like was discussed in https://github.com/pydata/xarray/pull/8872#issuecomment-2030406780

Comment (dcherian) [reply xarray/tests/test_dataset.py:3438]: Yes, I think that if a user explicitly writes `ds.expand_dims("x", create_index=True)` and we don't create an index, we should either warn or error.

In this case, it seems OK to warn and ask the user to call `set_coords("x")` first. We want to encourage that behaviour anyway.

I have not thought about the global option yet, but usually we handle that by setting `None` as the default value and then harmonizing with the global setting.

Comment (dcherian) [reply xarray/tests/test_dataset.py:3438]: I'm also OK punting that to a future PR. The current behaviour is pretty broken already.

Comment (TomNicholas) [reply xarray/tests/test_dataset.py]: Added a `UserWarning` in https://github.com/pydata/xarray/pull/8960/commits/7e8f895c2746e2a75979b5675967b434aef09ee6.

**Review by dcherian — APPROVED** [review]

**Pull request merged**